### PR TITLE
8234534: Simplify CardTable code after CMS removal 

### PR DESCRIPTION
**Pull request by tschatzl**
Hi,

  can I have reviews for this cleanup that removes CMS specific code from `CardTable/CardTableRS`?

Note that there is still this "conc_scan" parameter passed to the card table that affects barrier code generation, for some reason also G1 barrier code generation although it should not as `G1CardTable::scanned_concurrently()` only used for the "normal" card table. Initial attempts showed that removing this is not straightforward, causing crashes and so I left it out for [JDK-8250941](https://bugs.openjdk.java.net/browse/JDK-8260941) so that this change is solely about removing unused code.

Testing: tier1-4, some tier1-5 runs earlier (before some removal of hunks for files only containing copyright updates or newline changes)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8234534](https://bugs.openjdk.java.net/browse/JDK-8234534): Simplify CardTable code after CMS removal

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author) ⚠️ Review applies to 5aa23d74521f4bbc941d98fe85f0b21cbedfe622
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to 5aa23d74521f4bbc941d98fe85f0b21cbedfe622

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2354/head:pull/2354`
`$ git checkout pull/2354`

### CODE REVIEW

**Comment by tschatzl** [comment]
(latest tier1-4 testing still stuck on linux-aarch64, but everything else passed. I think there is no particular aarch64 specific change in there...)

File: src/hotspot/share/gc/shared/cardTableRS.cpp [hunk 1]
```diff
@@ -480,58 +437,12 @@ void CardTableRS::verify() {
...
+  CardTable(whole_heap, scanned_concurrently) { }
 
-CardTableRS::~CardTableRS() {
-  FREE_C_HEAP_ARRAY(CardValue, _last_cur_val_in_gen);
-  FREE_C_HEAP_ARRAY(CardArr, _lowest_non_clean);
-  FREE_C_HEAP_ARRAY(size_t, _lowest_non_clean_chunk_size);
-  FREE_C_HEAP_ARRAY(uintptr_t, _lowest_non_clean_base_chunk_index);
-  FREE_C_HEAP_ARRAY(int, _last_LNC_resizing_collection);
-}
+CardTableRS::~CardTableRS() { }
```
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/cardTableRS.cpp:442]: Now that it's empty, is it possible to remove it completely?

File: src/hotspot/share/gc/shared/cardTableRS.hpp [hunk 2]
```diff
@@ -92,18 +52,9 @@ class CardTableRS: public CardTable {
 
   virtual void verify_used_region_at_save_marks(Space* sp) const NOT_DEBUG_RETURN;
 
-  // Override.
-  void prepare_for_younger_refs_iterate(bool parallel);
-
-  // Card table entries are cleared before application;
-  void at_younger_refs_iterate();
-
   void inline_write_ref_field_gc(void* field, oop new_val) {
```
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/cardTableRS.hpp:55]: It seems that the arg `new_val` is not used. Maybe remove it or add a comment saying it's an intentional omission.

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/shared/cardTableRS.cpp [hunk 3]
```diff
@@ -38,53 +38,10 @@
...
-}
-
 inline bool ClearNoncleanCardWrapper::clear_card(CardValue* entry) {
   CardValue entry_val = *entry;
-  assert(entry_val != CardTableRS::clean_card_val(),
-         "We shouldn't be looking at clean cards, and this should "
-         "be the only place they get cleaned.");
-  assert(entry_val != CardTableRS::cur_youngergen_and_prev_nonclean_card,
-         "This should be possible in the sequential case.");
+  assert(entry_val == CardTableRS::dirty_card_val(),
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/cardTableRS.cpp:43]: Consider eliminating `entry_val` - just use `*entry` in the assert.

**Review by kimbarrett — APPROVED** [review]
Looks good to me, with the one minor nit I commented on and Albert's suggestions.

**Comment by tschatzl** [comment]
All fixed as suggested. Still compiles.

**Comment by tschatzl** [comment]
Thanks @kimbarrett @albertnetymk for your reviews.

/integrate

**Pull request closed without merging**